### ISSUE TITLE
PyTorch Profiling Tools

### ISSUE THREAD

**Issue by jiweiqi** [DENG-MIT/reactorch#12]
Start an issue to list available PyTorch profiling tools

**Comment by jiweiqi** [comment]
HI, @taosong1999 You may provide an example line of code to show how to profile the solution_test.py and how to interpret the profile results here.

**Comment by jiweiqi** [comment]
[Line by Line Profiling in iPython](https://jakevdp.github.io/PythonDataScienceHandbook/01.07-timing-and-profiling.html)

**Comment by WeilunQiu** [comment]
Hi, all. Here's a performance test on the reaction rate constants func (forward, equlibrium and reverse) and wdot func, in which I run these functions for several times (see x-axis). Note that the path of the mechanism file might need to be changed according to where you put the script.

Results:
(The unit of y-axis should be second, of course. )
![Performance_test](https://user-images.githubusercontent.com/41820879/85950552-06478100-b990-11ea-8779-f4f43a2c6c4a.png)

Codes:
```Python
from time import perf_counter

import cantera as ct
import matplotlib.pyplot as plt
import numpy as np
import torch

import reactorch as rt

cpu = torch.device('cpu')

cuda = torch.device('cuda:0')

device = cpu

mech_yaml = '../../data/gri30.yaml'
composition = "CH4:0.5, O2:1.0, N2:3.76"

sol = rt.Solution(mech_yaml=mech_yaml, device=device)

gas = sol.gas
gas.TPX = 950, 20 * ct.one_atm, composition

r = ct.IdealGasReactor(gas)
sim = ct.ReactorNet([r])

time = 0.0
t_end = 10
idt = 0
states = ct.SolutionArray(gas, extra=['t'])
T0 = gas.T

print('%10s %10s %10s %14s' % ('t [s]', 'T [K]', 'P [atm]', 'u [J/kg]'))

count = 0
while sim.time < t_end:

    sim.step()

    states.append(r.thermo.state, t=time)

    if r.thermo.T > T0 + 600 and idt < 1e-10:
        idt = sim.time

    if idt > 1e-10 and sim.time > 4 * idt:
        break

print('%10.3e %10.3f %10.3f %14.6e' % (sim.time,
                                       r.T,
                                       r.thermo.P / ct.one_atm,
                                       r.thermo.u))

print('idt = {:.2e} [s] number of points {}'.format(idt, states.t.shape[0]))

TP = torch.stack((torch.Tensor(states.T), torch.Tensor(states.P)), dim=-1)
Y = torch.Tensor(states.Y)
TPY = torch.cat([TP, Y], dim=-1).to(device)

t0_start = perf_counter()

sol.set_states(TPY)

t1_stop = perf_counter()
print('sol set_states time spent {:.1e} [s]'.format(t1_stop - t0_start))

# tesing reaction rate func
def check_reaction_rate(num):
       for i in range(0,num):
              sol.forward_rate_constants_func()
              sol.equilibrium_constants_func()
              sol.reverse_rate_constants_func()
       
# tesing wdot calc func
def check_wdot(num):
       for i in range(0,num):
              sol.wdot_func()

test_n_iter = np.array([100,150,200,250,300,350,400,450,500])
time_cost1 = np.zeros_like(test_n_iter)
time_cost2 = np.zeros_like(test_n_iter)

for j in range(0,9):
       print('#Test {:d}'.format(j))
             
       t1_stop = perf_counter()
       
       check_reaction_rate(test_n_iter[j])
       t2_stop = perf_counter()
       t_cost1 = t2_stop - t1_stop
       print('sol check_reaction_rate time spent {:.1e} [s]'.format(t_cost1))      
       
       check_wdot(test_n_iter[j])
       t3_stop = perf_counter()
       t_cost2 = t3_stop - t2_stop
       print('sol check_wdot time spent {:.1e} [s]'.format(t_cost2))
       
       time_cost1[j] = t_cost1 
       time_cost2[j] = t_cost2
       
plt.plot(test_n_iter,time_cost1)
plt.plot(test_n_iter,time_cost2)
plt.xlabel('Number of iterations')
plt.ylabel('Time cost')
plt.legend(['Reaction rate cost','Wdot cost'])
plt.savefig('Performance_test.png',dpi=600)
```
